lib/network/level-net/index.js

```
@@ -0,0 +1,283 @@
+
+/**
+ * @brief create a server object that serves remote LevelDB requests
+ * through socket.io events
+ *
+ * @param {Object} db The root LevelDB database object to expose to
+ * remote clients
+ * @param {Object} options currently supported options are:
+ * - streamMaxPendingAck {Number} max number of in-flight output
+ *   stream packets sent to the client without an ack received yet
+ * - streamAckTimeoutMs {Number} timeout for receiving an ack after an
+ *   output stream packet is sent to the client
+ * @return {Object} a server object, not yet listening on a TCP port
+ * (call listen() next for this)
+ */
+module.exports.createServer = function LevelServer(db, options = {}) {
+    const httpServer = http.createServer();
+    const ioServer = io(httpServer);
+
+    ioServer.initMetadataService = function initMetadataService(baseNs = '') {
+        assert(baseNs === '' || baseNs.startsWith('/'));
+
+        const mdSock = this.of(`${baseNs}/metadata`);
+        mdSock.on('connection', conn => {
+            function openSub(path) {
+                let subDb = db;
+                path.forEach(pathItem => {
+                    subDb = subDb.sublevel(pathItem);
+                });
+                return subDb;
+            }
+
+            const mdStreams = sioStream(conn,
+                                        options.streamMaxPendingAck,
+                                        options.streamAckTimeoutMs);
+            conn.on('error', () => {});
+            conn.on('call', (remoteCall, args, cb) => {
+                if (dbAsyncCommandNames.includes(remoteCall)) {
+                    try {
+                        const path = args.pop();
+                        args.push((err, data) => {
+                            cb(serializeError(err),
+                               mdStreams.encodeStreams(data));
+                        });
+
+                        const subDb = openSub(path);
+                        subDb[remoteCall].apply(subDb, args);
```

I don't think that subDb[remoteCall] may throw (I hope not, it ought to call the cb instead). I don't think that openSub throws either (via `db.sublevel()`) but I'm not 100% sure. Though if it does, I think there's something going seriously wrong so we could let the exception bubble up and crash the process. (so the try/catch may be removed here). I don't have a strong opinion on keeping or removing it here.